cloud/WFLY-Operator-100-report-txn-heuristic-state-better.adoc

```diff
@@ -0,0 +1,106 @@
+= Transaction heuristic state should be reported in a better way
+:author:            Ondrej Chaloupka
+:email:             ochaloup@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+:keywords:          openshift,transactions,recovery
+
+
+== Overview
+
+If a pod is scale down then all transactions need to be finished before the pod is terminated.
+It's responsibility of WildFly Operator to maintain the transaction consistency in this manner.
+The WildFly operator runs transaction recovery for his purpose to clean the transaction log.
+During the transaction recovery process there are some cases which are not presented to the user
+in a good manner. These need to be enhanced. Then a developer mode where transaction recovery
+is not run at all needs to be offered in the WildFly Operator functionality.
+
+== Issue Metadata
+
+=== Issue
+
+* https://github.com/wildfly/wildfly-operator/issues/100
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/CLOUD-2262[CLOUD-2262]
+
+=== Dev Contacts
+
+* mailto:ochaloup@redhat.com[{author}]
+
+=== QE Contacts
+
+* ?
+
+=== Testing By
+
+[x] Engineering
+[ ] QE
+
+=== Affected Projects or Components
+
+* WildFly Operator
+
+=== Other Interested Projects
+
+* Transactions
+
+== Description of the feature
+
+When transaction scale down process reviews the state of transactions in the pods it checks if there are no unfinished prepared transactions.
+If there are found some then it does not permit the pod being terminated and the pod is marked with the _state_ `SCALING_DOWN_RECOVERY_DIRTY`.
+It's marked as dirty until all transactions are safely recovered.
```

Comment:
Is pod fully working in dirty state? Is load balancer forwarding traffic to such pod is it handling requests? 
What happens when I scale up to origin number of pods again? I understand dirty state of pod is not changed. 
Can I scale up to bigger number of pods?

I am thinking of usecase where Autoscaler is involved based on current cluster utilization. So pods get scaledown and scaleup randomly. How it will happen in this situation?

Reply:
This summarizes how the current functionality is already working.
* When in dirty state the pod is not capable to start a new transaction. But otherwise it's working. There is a issue to manage that in a better way here: https://issues.redhat.com/browse/EAP7-1355
* No, load balancer is not forwarding the traffic to the pod
* The pod should be moved to the state ACTIVE. The pod is not dirty anymore as it's normally operating and normally operating pod may process recovery.
* Yes, you can scale up as you wish. The processing of scaledown does not touch it

All these situations should be part of the eap xpaas testsuite.

I assume the scaledown and scaleup processing in a faster pace should be working fine. Just it could happen that the pod won't be really never scaleddown as scaling down really takes time. That needs to be checked.